crates/ra_hir/src/expr.rs

```diff
@@ -613,6 +615,10 @@ impl ExprCollector {
                         .map(|arm| MatchArm {
                             pats: arm.pats().map(|p| self.collect_pat(p)).collect(),
                             expr: self.collect_expr_opt(arm.expr()),
+                            guard: arm.guard().map(|guard| {
+                                let e = guard.expr().expect("every guard should have an expr");
```

Comment:
is that so? what about
```
match foo {
  x if => {}
}
```
?

and even if that somehow currently results in a different parse tree, I think it's more robust to just handle the case of a missing expr -- setting the match arm to `None` or `Some` with `Expr::Missing` would both be fine, I think

Reply:
rustc has the luxury of bailing out on parse errors; we don't want to crash on invalid syntax and ideally want to still provide completions etc. :)